From f821d30d3a8313267e8cba5e703dfb10576e65fa Mon Sep 17 00:00:00 2001
From: Vlad Yasevich <vyasevic@redhat.com>
Date: Tue, 16 Dec 2014 14:23:36 -0500
Subject: [net] bridge: Simplify pvid checks

Message-id: <1418739817-22923-7-git-send-email-vyasevic@redhat.com>
Patchwork-id: 102094
O-Subject: [RHEL7 PATCH 6/7] bridge: Simplify pvid checks.
Bugzilla: 1164653
RH-Acked-by: Michael S. Tsirkin <mst@redhat.com>
RH-Acked-by: Marcelo Leitner <mleitner@redhat.com>
RH-Acked-by: Jiri Benc <jbenc@redhat.com>

BZ: 1164653
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=8371670

Upstream commit 3df6bf45ec008942f16f1814123c4bdebcf50741
Author: Vlad Yasevich <vyasevich@gmail.com>
Date:   Fri Oct 3 11:29:17 2014 -0400

    bridge: Simplify pvid checks.

    Currently, if the pvid is not set, we return an illegal vlan value
    even though the pvid value is set to 0.  Since pvid of 0 is currently
    invalid, just return 0 instead.  This makes the current and future
    checks simpler.

    Signed-off-by: Vladislav Yasevich <vyasevic@redhat.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Vladislav Yasevich <vyasevic@redhat.com>

Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/bridge/br_private.h b/net/bridge/br_private.h
index 397b4ed..beb5e52 100644
--- a/net/bridge/br_private.h
+++ b/net/bridge/br_private.h
@@ -657,11 +657,8 @@ static inline int br_vlan_get_tag(const struct sk_buff *skb, u16 *vid)
 
 static inline u16 br_get_pvid(const struct net_port_vlans *v)
 {
- /* Return just the VID if it is set, or VLAN_N_VID (invalid vid) if
-  * vid wasn't set
-  */
  smp_rmb();
- return v->pvid ?: VLAN_N_VID;
+ return v->pvid;
 }
 
 static inline int br_vlan_enabled(struct net_bridge *br)
@@ -756,7 +753,7 @@ static inline u16 br_vlan_get_tag(const struct sk_buff *skb, u16 *tag)
 }
 static inline u16 br_get_pvid(const struct net_port_vlans *v)
 {
- return VLAN_N_VID; /* Returns invalid vid */
+ return 0;
 }
 
 static inline int br_vlan_enabled(struct net_bridge *br)
diff --git a/net/bridge/br_vlan.c b/net/bridge/br_vlan.c
index 4181e48..1f1bada 100644
--- a/net/bridge/br_vlan.c
+++ b/net/bridge/br_vlan.c
@@ -217,7 +217,7 @@ bool br_allowed_ingress(struct net_bridge *br, struct net_port_vlans *v,
    * See if pvid is set on this port.  That tells us which
    * vlan untagged or priority-tagged traffic belongs to.
    */
-  if (pvid == VLAN_N_VID)
+  if (!pvid)
    goto drop;
 
   /* PVID is set on this port.  Any untagged or priority-tagged
@@ -284,7 +284,7 @@ bool br_should_learn(struct net_bridge_port *p, struct sk_buff *skb, u16 *vid)
 
  if (!*vid) {
   *vid = br_get_pvid(v);
-  if (*vid == VLAN_N_VID)
+  if (!*vid)
    return false;
 
   return true;
-- 
1.7.1